Self register domains restriction

MarieVerdonck opened the pull request:
## References
* Fixes https://github.com/DSpace/DSpace/issues/3281
* Fixes https://github.com/DSpace/DSpace/issues/3272
* Angular PR: https://github.com/DSpace/dspace-angular/pull/2016
* Contract changes: https://github.com/DSpace/RestContract/pull/211

## Description

Checks the config `authentication-password.domain.valid` for valid domains before new e-mail registration. If the e-mail doesn't end in any of the values in this config, registration is not allowed.

Also introduces new required query param to the `/api/eperson/registrations` endpoint to contain either 'forgot' or 'register'. Previously it was assumed that if the attached e-mail address wasn't registered that it was to request a register e-mail and vice versa, regardless from which form in UI (forgot/register) it came. But with the new domain restriction this means you could use the endpoint to determine which e-mail addresses are already in use. Examples:

- `test@email.com` user exists (only `@university.com` in allowed domains list) => Register form with `test@email.com` won't fail despite it being stated this is not in allowed domains list => Now know this e-mail already in use
- `test2@email.com` user doesn't exist (only `@university.com` in allowed domains list) => Forgot form with `test2@email.com` => fails because user doesn't exist & register fails because not allowed domain => Now know this e-mail is not already in use

## Instructions for Reviewers

See Angular PR: https://github.com/DSpace/dspace-angular/pull/2016

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change; https://github.com/DSpace/DSpace/pull/8619
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue committed:
Update to Spring v5.2.20. Requires also explicitly specifying version of spring-context-support to avoid dependency convergence issues.

tdonohue committed:
Change to snapshot version

tdonohue committed:
[maven-release-plugin] prepare release dspace-7.2.1

tdonohue committed:
[maven-release-plugin] prepare for next development iteration

joost-atmire committed:
[task 97298] implemented check for allowed domains during registration'

joost-atmire committed:
Merge remote-tracking branch 'upstream/main' into w2p-97298_issue-3281_self-register-issue-main

joost-atmire committed:
[task 97298] only check for allowed domains on register

joost-atmire committed:
Merge branch 'w2p-97298_issue-3281_self-register-issue-7.2' into w2p-97298_issue-3281_self-register-issue-main

MarieVerdonck committed:
Merge remote-tracking branch 'origin/main' into w2p-97298_issue-3281_self-register-issue-main

MarieVerdonck committed:
97298: #3281 Self-register - type param

MarieVerdonck committed:
Merge branch 'w2p-97298_issue-3281_self-register-issue-7.2' into w2p-97298_issue-3281_self-register-issue-main

MarieVerdonck committed:
97298: #3281 Self-register - fix, needs existent user for forgot

MarieVerdonck committed:
97298: Self-register - test fixes

MarieVerdonck committed:
Merge branch 'w2p-97298_issue-3281_self-register-issue-7.2' into w2p-97298_issue-3281_self-register-issue-main

MarieVerdonck committed:
97298: Self-register - test fixes

tdonohue reviewed (commented):
@MarieVerdonck : Overall this looks good and it works well based on my testing (I also tested that forgot password still works fully, and it does).  I just have a small inline question below about a POM change that doesn't make sense to me.

I'm basically a +1, but I'm not sure I understand why the POM change is needed as none of the new code in this PR seems to use it.

tdonohue commented on dspace/config/modules/authentication-password.cfg:
```diff
@@ -6,6 +6,9 @@
 #
 
+# self-registration can be disabled completely by setting the user.registration property to false
+# user.registration = false
```
Forgot to add a comment here.  While it's good to add this setting in our configs, I noticed it doesn't seem to work.  But, I'm assuming this doesn't work because #3272 isn't fixed yet?

MarieVerdonck replied:
No, this does work, I think you misread `user.registration = false` (commented out) mean you can't register, setting it to true (the default) would make it so registration doesn't work

- See https://github.com/DSpace/dspace-angular/pull/2016: With config set (in `local.cfg`): `user.registration=false` => `{ui.url}/register` not accessible (redirects login or forbidden if logged in) & button to it in login dropdown not present

tdonohue replied:
@MarieVerdonck : You were correct. I completely misread this config :)  tested again today and it works exactly as expected. When set to `user.registration=false`, I see registration is disabled/turned off, and I no longer see that option in the UI level.  So, this proves to me that this PR *fixes* #3272.

MarieVerdonck committed:
97061: type request param name change to avoid confusion with rest object type &
error codes endpoint changes + IT:
- missing/wrong query param accountRequestType => 400 bad request
- email not valid domain => 422 unprocessable entity

MarieVerdonck committed:
Merge branch 'w2p-97298_issue-3281_self-register-issue-7.2' into w2p-97298_issue-3281_self-register-issue-main

# Conflicts:
#	dspace-server-webapp/src/test/java/org/dspace/app/rest/RegistrationRestRepositoryIT.java

MarieVerdonck commented:
Changes because of requested [[contract](https://github.com/DSpace/RestContract/pull/211)](https://github.com/DSpace/RestContract/pull/211) changes:

- `type` query param ('forgot' or 'register' value) changed to `accountRequestType` to avoid confusion with rest object type field
- Missing `accountRequestType` query param or value in it that is not forgot nor register will result in 400 bad request error
- E-mail address given now not in domain will given 422 unprocessable entity error

tdonohue commented on pom.xml:
```diff
@@ -1230,6 +1230,12 @@
+                <groupId>org.springframework</groupId>
+                <artifactId>spring-context-support</artifactId>
+                <version>${spring.version}</version>
+            </dependency>
```
@MarieVerdonck : In that case, I think this can be safely removed from this PR.  It's not necessary on `main` anymore, so there's no need to re-add it in this PR.

MarieVerdonck committed:
PR #8619: Removed dependency spring-context-support, not needed in main

tdonohue reviewed (changes requested):
@MarieVerdonck : This looks great now & I verified it works.  However, an accidental change was made to `rest.cfg` which needs to be undone.  See inline below.

abollini reviewed (approved):
thanks @MarieVerdonck it looks good to me as well. Just need to revert the accidental changes noted by @tdonohue

MarieVerdonck committed:
Reinstate config accidentally deleted, added in 30f923a080aa0d927036e0eae609c3bf1a273d35 (#8064) - enables RSS/Atom feeds

tdonohue reviewed (approved):
👍 Thanks @MarieVerdonck !  This looks good now.  I'll wait to merge it though until the frontend PR is ready/approved (https://github.com/DSpace/dspace-angular/pull/2016), just in case other minor changes are necessary on the backend in finalizing the frontend.
